File: MotoActions/src/org/lineageos/settings/device/actions/DozePulse.java
/*
 * SPDX-FileCopyrightText: 2015 The CyanogenMod Project
 * SPDX-FileCopyrightText: The LineageOS Project
 * SPDX-License-Identifier: Apache-2.0
 */

package org.lineageos.settings.device.actions;

import android.content.Context;
import android.content.Intent;
import android.hardware.Sensor;
import android.hardware.SensorEvent;
import android.hardware.SensorEventListener;
import android.os.UserHandle;
import android.util.Log;

import org.lineageos.settings.device.MotoActionsSettings;
import org.lineageos.settings.device.SensorHelper;

public class DozePulse {
    private static final String TAG = "MotoActions-DozePulse";

    // Maximum time for the hand to cover the sensor: 1s
    private static final long HANDWAVE_MAX_DELTA_NS = 1000L * 1000 * 1000;

    // Minimum time until the device is considered to have been in the pocket: 5s
    private static final long POCKET_MIN_DELTA_NS = 5000L * 1000 * 1000;

    // Minimum time between doze pulses: 1.5s
    private static final int DELAY_BETWEEN_DOZES_IN_MS = 1500;

    private final Context mContext;
    private final MotoActionsSettings mMotoActionsSettings;
    private final SensorHelper mSensorHelper;
    private final Sensor mFlatUpSensor;
    private final Sensor mStowSensor;

    private boolean mEnabled;
    private boolean mIsFlatUp;
    private boolean mLastFlatUp;
    private boolean mIsStowed;
    private boolean mLastStowed;
    private long mLastStowedTime;
    private long mLastDoze;

    public DozePulse(MotoActionsSettings MotoActionsSettings, Context context,
                        SensorHelper sensorHelper) {
        mContext = context;
        mMotoActionsSettings = MotoActionsSettings;
        mSensorHelper = sensorHelper;

        mFlatUpSensor = sensorHelper.getFlatUpSensor();
        mStowSensor = sensorHelper.getStowSensor();
    }

    public void screenTurnedOn() {
        if (mEnabled) {
            Log.d(TAG, "Disabling");
            mSensorHelper.unregisterListener(mFlatUpListener);
            mSensorHelper.unregisterListener(mStowListener);
            mEnabled = false;
        }
    }

    public void screenTurnedOff() {
        if (!mEnabled) {
            Log.d(TAG, "Enabling");
            mSensorHelper.registerListener(mFlatUpSensor, mFlatUpListener);
            mSensorHelper.registerListener(mStowSensor, mStowListener);
            mEnabled = true;
        }
        mLastDoze = System.currentTimeMillis();
    }

    private final SensorEventListener mFlatUpListener = new SensorEventListener() {
        @Override
        public synchronized void onSensorChanged(SensorEvent event) {
            mIsFlatUp = (event.values[0] != 0);

            Log.d(TAG, "event: " + mIsFlatUp + " mLastFlatUp=" + mLastFlatUp + " mIsStowed=" +
                    mIsStowed);

            boolean canPulse = mMotoActionsSettings.isPickUpEnabled()
                    && (mLastFlatUp && !mIsFlatUp && !mIsStowed);
            sendDozePulse(canPulse);
            mLastFlatUp = mIsFlatUp;
        }

        @Override
        public void onAccuracyChanged(Sensor sensor, int accuracy) {
        }
    };

    private final SensorEventListener mStowListener = new SensorEventListener() {
        @Override
        public synchronized void onSensorChanged(SensorEvent event) {
            mIsStowed = (event.values[0] != 0);
            if (mIsStowed) {
                mLastStowedTime = event.timestamp;
            } else if (mLastStowed) {
                sendDozePulse(shouldPulse(event.timestamp));
            }
            mLastStowed = mIsStowed;
            Log.d(TAG, "event: " + mIsStowed);
        }

        private boolean shouldPulse(long timestamp) {
            long delta = timestamp - mLastStowedTime;

            boolean irWakeupEnabled = mMotoActionsSettings.isIrWakeupEnabled();
            boolean pocketGestureEnabled = mMotoActionsSettings.isPocketGestureEnabled();

           if (irWakeupEnabled && pocketGestureEnabled) {
                return true;
            } else if (irWakeupEnabled) {
                return delta < HANDWAVE_MAX_DELTA_NS;
            } else if (pocketGestureEnabled) {
                return delta >= POCKET_MIN_DELTA_NS;
            }
            return false;
        }

        @Override
        public void onAccuracyChanged(Sensor sensor, int accuracy) {
        }
    };

    public void sendDozePulse(boolean canPulse) {
        if (!canPulse) {
            return;
        }

        long now = System.currentTimeMillis();
        if (now - mLastDoze > DELAY_BETWEEN_DOZES_IN_MS) {
            mLastDoze = now;
            Log.d(TAG, "Launching doze pulse");
            Intent pulseIntent = new Intent("com.android.systemui.doze.pulse");
            mContext.sendBroadcastAsUser(pulseIntent, UserHandle.CURRENT);
        }
    }
}

File: packages/MotoActions/src/org/lineageos/settings/device/actions/DozePulse.kt
/*
 * SPDX-FileCopyrightText: 2015 The CyanogenMod Project
 * SPDX-FileCopyrightText: The LineageOS Project
 * SPDX-License-Identifier: Apache-2.0
 */

package org.lineageos.settings.device.actions

import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.os.PowerManager
import android.os.PowerManager.WakeLock
import android.os.UserHandle
import android.util.Log
import org.lineageos.settings.device.MotoActionsSettings
import org.lineageos.settings.device.MotoActionsSettings.GESTURE_IR_WAKEUP_KEY
import org.lineageos.settings.device.MotoActionsSettings.GESTURE_PICK_UP_KEY
import org.lineageos.settings.device.MotoActionsSettings.GESTURE_POCKET_KEY
import org.lineageos.settings.device.SensorHelper
import org.lineageos.settings.device.SensorHelper.Companion.SENSOR_TYPE_MMI_FLAT_UP
import org.lineageos.settings.device.SensorHelper.Companion.SENSOR_TYPE_MMI_STOW

class DozePulse(
    private val context: Context,
    private val sharedPreferences: SharedPreferences,
    private val sensorHelper: SensorHelper,
) : UpdatedStateNotifier {

    private val powerManager: PowerManager = context.getSystemService(PowerManager::class.java)
    private val wakeLock: WakeLock =
        powerManager.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK, "$TAG:WakeLock")
    private val flatUpSensor: Sensor = sensorHelper.getSensor(SENSOR_TYPE_MMI_FLAT_UP)!!
    private val stowSensor: Sensor = sensorHelper.getSensor(SENSOR_TYPE_MMI_STOW)!!

    private var isStowed = false
    private var lastStowed = false
    private var lastStowedTime = 0L

    private var lastFlatUp = false

    private var lastDozeTime: Long = 0L

    fun onScreenStateChanged(screenOn: Boolean) {
        if (!wakeLock.isHeld) {
            if (screenOn) {
                wakeLock.acquire()
            } else {
                wakeLock.release()
            }
        }

        if (screenOn) {
            Log.d(TAG, "Disabling")
            sensorHelper.unregisterListener(flatUpListener)
            sensorHelper.unregisterListener(stowListener)
        } else {
            Log.d(TAG, "Enabling")
            sensorHelper.registerListener(flatUpSensor, flatUpListener)
            sensorHelper.registerListener(stowSensor, stowListener)
            lastDozeTime = System.currentTimeMillis()
        }
    }

    @Synchronized
    override fun updateState() {
        onScreenStateChanged(powerManager.isInteractive)
    }

    private val flatUpListener =
        object : SensorEventListener {
            @Synchronized
            override fun onSensorChanged(event: SensorEvent) {
                val thisFlatUp = event.values[0] != 0f
                val pickUpGestureEnabled = sharedPreferences.getBoolean(GESTURE_PICK_UP_KEY, true)
                if (pickUpGestureEnabled && lastFlatUp && !thisFlatUp && !isStowed) {
                    onStartPulse()
                }
                lastFlatUp = thisFlatUp
                Log.d(TAG, "event: $thisFlatUp")
            }

            override fun onAccuracyChanged(sensor: Sensor, accuracy: Int) {}
        }

    private val stowListener =
        object : SensorEventListener {
            @Synchronized
            override fun onSensorChanged(event: SensorEvent) {
                isStowed = event.values[0] != 0f
                if (isStowed) {
                    lastStowedTime = event.timestamp
                } else if (lastStowed && shouldPulse(event.timestamp)) {
                    onStartPulse()
                }
                lastStowed = isStowed
                Log.d(TAG, "event: $isStowed")
            }

            private fun shouldPulse(timestamp: Long): Boolean {
                val delta = timestamp - lastStowedTime

                val irWakeupEnabled = sharedPreferences.getBoolean(GESTURE_IR_WAKEUP_KEY, true)
                val pocketGestureEnabled = sharedPreferences.getBoolean(GESTURE_POCKET_KEY, true)

                return when {
                    irWakeupEnabled && pocketGestureEnabled -> true
                    irWakeupEnabled -> delta < HANDWAVE_MAX_DELTA_NS
                    pocketGestureEnabled -> delta >= POCKET_MIN_DELTA_NS
                    else -> false
                }
            }

            override fun onAccuracyChanged(sensor: Sensor, accuracy: Int) {}
        }

    private fun onStartPulse() {
        if (
            !MotoActionsSettings.isDozeEnabled(context) ||
                MotoActionsSettings.isAlwaysOnEnabled(context)
        ) {
            return
        }
        val now = System.currentTimeMillis()
        if (now - lastDozeTime > DELAY_BETWEEN_DOZES_IN_MS) {
            lastDozeTime = now
            context.sendBroadcastAsUser(
                Intent("com.android.systemui.doze.pulse"),
                UserHandle.CURRENT,
            )
        }
    }

    companion object {
        private const val TAG = "DozePulse"
        private const val DELAY_BETWEEN_DOZES_IN_MS = 1500
        private const val HANDWAVE_MAX_DELTA_NS = 1000L * 1000 * 1000
        private const val POCKET_MIN_DELTA_NS = 5000L * 1000 * 1000
    }
}
